[visionOS] MediaPlayer changes while in external playback get stuck in the wrong rendering mode.

rsfuller opened the pull request:
#### ad09364a059eb2fab58744bcba57532e565447c9
<pre>
[visionOS] MediaPlayer changes while in external playback get stuck in the wrong rendering mode.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293755">https://bugs.webkit.org/show_bug.cgi?id=293755</a>
<a href="https://rdar.apple.com/151717489">rdar://151717489</a>

Reviewed by Andy Estes.

When an HTMLVideoElement is in external playback (a custom LMK mode) and the media
player changes, the new media player does not receive the inFullscreenOrPiP state
correctly. Thus, when exiting external playback, the media element short circuits
the call to the media player&apos;s private interface, which means the rendering mode is
not updated from entity to layer based.

The new media player already does get its inFullscreenOrPip value copied over from
the fullscreen state of the element, but this does not currently capture whether
the element is in external playback -- which also should be considered to be in
fullscreenOrPip.

To fix, adjust the calculation when creating a new media player to consider external
state, which requires creating a variable on HTMLVideoElement to store the state.

Also, synchronize some existing external playback code under the LINEAR_MEDIA_PLAYER
flag instead of just visionOS, as that is more technically correct as a requirement
and should have a consistent guard around all external playback code.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::isInFullscreenOrPictureInPicture const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::didEnterExternalPlayback):
(WebCore::HTMLVideoElement::didExitExternalPlayback):
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.messages.in:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:

Canonical link: <a href="https://commits.webkit.org/295573@main">https://commits.webkit.org/295573@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3917ae688a0510b0682008da647d305327ff8e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105496 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15635 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110696 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/56152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107537 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25674 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33748 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80123 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/56152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108502 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95211 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60432 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19762 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13301 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55533 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89492 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13342 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113471 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32638 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89205 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33001 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91440 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88864 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22657 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33734 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/11538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/28094 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32564 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/37976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32319 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35666 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33907 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -8076,7 +8087,7 @@ void HTMLMediaElement::createMediaPlayer() WTF_IGNORES_THREAD_SAFETY_ANALYSIS
     player->setPageIsVisible(!m_elementIsHidden);
     player->setVisibleInViewport(isVisibleInViewport());
-    player->setInFullscreenOrPictureInPicture(isFullscreen());
+    player->setInFullscreenOrPictureInPicture(isInFullscreenOrPictureInPicture());
```
This seems OK for now, but have you considered making external playback be a new kind of `VideoFullscreenMode` alongside `VideoFullscreenModeStandard`, `VideoFullscreenModePictureInPicture`, etc.?

rsfuller committed:
[visionOS] MediaPlayer changes while in external playback get stuck in the wrong rendering mode.
https://bugs.webkit.org/show_bug.cgi?id=293755
rdar://151717489

Reviewed by Andy Estes.

When an HTMLVideoElement is in external playback (a custom LMK mode) and the media
player changes, the new media player does not receive the inFullscreenOrPiP state
correctly. Thus, when exiting external playback, the media element short circuits
the call to the media player's private interface, which means the rendering mode is
not updated from entity to layer based.

The new media player already does get its inFullscreenOrPip value copied over from
the fullscreen state of the element, but this does not currently capture whether
the element is in external playback -- which also should be considered to be in
fullscreenOrPip.

To fix, adjust the calculation when creating a new media player to consider external
state, which requires creating a variable on HTMLVideoElement to store the state.

Also, synchronize some existing external playback code under the LINEAR_MEDIA_PLAYER
flag instead of just visionOS, as that is more technically correct as a requirement
and should have a consistent guard around all external playback code.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::isInFullscreenOrPictureInPicture const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::didEnterExternalPlayback):
(WebCore::HTMLVideoElement::didExitExternalPlayback):
* Source/WebCore/html/HTMLVideoElement.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.messages.in:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:

Canonical link: https://commits.webkit.org/295573@main
